python-package/lightgbm/basic.py

```diff
@@ -546,9 +551,8 @@ def _data_from_pandas(data, feature_name, categorical_feature, pandas_categorica
             raise ValueError("DataFrame.dtypes for data must be int, float or bool.\n"
                              "Did not expect the data types in the following fields: "
                              f"{bad_index_cols_str}")
-        data = data.values
-        if data.dtype != np.float32 and data.dtype != np.float64:
```

Comment:
This `if` statement was introduced with the aim to save memory in case of original data already have `float[32|64]` type: #2383. With new unconditional `data.astype(target_dtype).values` expression and [default argument `copy=True` of `astype()` function](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.astype.html), we are loosing that efficiency improvement. Am I wrong?

Reply:
The `np.ndarray.astype`s `copy` argument also defaults to `True` ([docs](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html#numpy.ndarray.astype)) so I think the behaviour is the same as it previously was. We could set `copy=False` here if we won't modify the values, and if the dtype matches (all of the columns are either f32 or f64) there won't be any copies.

Reply:
It'd be nice to test if any copies are made.

Reply:
> The `np.ndarray.astype`s `copy` argument also defaults to `True` ([docs](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html#numpy.ndarray.astype)) so I think the behaviour is the same as it previously was.

Previously, `np.ndarray.astype()` **wasn't executed at all** if dtype is `float` already. So, I think the behaviour is changed here.

Reply:
I added the `copy=False` argument and added a test to check that no copies are made for a single float dtype df in 98325e920def79a565b36fdcf7ede8a2b70c381e

Reply:
Note for other reviewers.
Casting to float[32|64] is done later here
https://github.com/microsoft/LightGBM/blob/e6a2f7162cf0bae4cbad6223c2700d74f0d67c4b/python-package/lightgbm/basic.py#L1595-L1598
for creating dataset and here
https://github.com/microsoft/LightGBM/blob/e6a2f7162cf0bae4cbad6223c2700d74f0d67c4b/python-package/lightgbm/basic.py#L849-L852
for predicting.

Reply:
> Looking at this a bit more closely I think I could add `np.float32` to ...

Sorry, didn't get how adding `np.float32` will help to avoid a loss of precision...

Reply:
Sorry. If we add floats to that function it will cast to float64, i.e.
```python
import numpy as np

np.find_common_type([np.int32, np.float32], [])  # float64
```
That also avoids a copy if we have like [int16, int32], the common dtype would be int32 so a copy would be made and then when casting to float another copy would be made. By including float32 in there we can cast to the target dtype only once.

Reply:
uhu, thanks for the explanation! I think it's great idea!

Reply:
added in 530828b